### ISSUE TITLE
Parallel runner not fixing any files

### ISSUE THREAD

**Issue by X-Coder264** [PHP-CS-Fixer/PHP-CS-Fixer#8014]
### Problem description

After enabling the parallel runner no files are being fixed anymore. It seems like CS fixer just finishes the process with the progress bar still being at 0%.

### Minimal reproducer

```php
$finder = (new PhpCsFixer\Finder())
    ->in(\dirname(__DIR__, 3) . '/src')
;

return (new PhpCsFixer\Config())
    //->setParallelConfig(PhpCsFixer\Runner\Parallel\ParallelConfigFactory::detect())
    ->setRules(
        [
            'nullable_type_declaration' => ['syntax' => 'union'],
        ],
    )
    ->setRiskyAllowed(true)
    ->setLineEnding("\n")
    ->setIndent('    ')
    ->setFinder($finder)
;
```

This is a simplified version of the config that I have that reproduces this problem and I've changed on purpose one method in my `src` folder to `abstract protected function getType(): ?DocumentType;`

After the fixer runs I'd expect to get `abstract protected function getType(): null|DocumentType;` as a result.
Running `php-cs-fixer fix --no-interaction --diff -vvv --using-cache=no` with this config without the parallel runner results in the file being fixed:

```
PHP CS Fixer 3.57.0 (5c90224) 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.2.18
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default from "./app/tools/php-cs-fixer/config.php".
 6703/6703 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) src/foo.php (nullable_type_declaration)
      ---------- begin diff ----------
--- /srv/app/src/foo.php
+++ /srv/app/src/foo.php
@@ -29,7 +29,7 @@
 
-    abstract protected function getType(): ?DocumentType;
+    abstract protected function getType(): null|DocumentType;


      ----------- end diff -----------


Fixed 1 of 6703 files in 12.973 seconds, 100.000 MB memory used
```

If I enable the parallel runner the file does not get fixed and the output is:

```
PHP CS Fixer 3.57.0 (5c90224) 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.2.18
Running analysis on 20 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "./app/tools/php-cs-fixer/config.php".
    0/6703 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fixed 0 of 6703 files in 0.530 seconds, 78.000 MB memory used
```

### Fixer version

3.57.0

### PHP version

8.2.18

### How do you run Fixer?

Composer Shim package (binary only)

### Contribution Checks

- [X] I have verified if this problem was already reported
- [X] I am familiar with ["Feature or bug?"](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst)

**Comment by Wirone** [comment]
@X-Coder264 could it be related to #8013, or you don't use `--path-mode=intersection`?

**Comment by X-Coder264** [comment]
I don't use `--path-mode=intersection`.

**Comment by keradus** [comment]
I will assume 3.57.1 is not fixing the issue.

Can you provide similar reproduce method to following ?
```
cd $(mktemp -d)
git clone https://github.com/verfriemelt-dot-org/wrapped . ## your repo
git checkout dae18e31aa003c818abe359f0e18378f89e19e12 ## your commit/branch
composer install
php-cs-fixer fix -vvv --using-cache=no --path-mode=intersection -- .php-cs-fixer.dist.php ## your command
```

**Comment by X-Coder264** [comment]
@keradus You assumed correctly, 3.57.1 is not fixing the issue for me.

The weird thing is I've tried the parallel runner on another project (that one doesn't use `php-cs-fixer/shim` though, it uses `friendsofphp/php-cs-fixer`) and it's working just fine there so I'm really confused as to what could be the problem here.

I'll remove the shim and install the non shim version tomorrow and see if that changes anything. If it doesn't I'll try to see what's going on with Xdebug if I'll have the time.

**Comment by X-Coder264** [comment]
Ok, I found the difference between the config of those two projects.

In the project in which the parallel runner does not work at all there are these two lines at the top of the CS fixer config file:

```php
require \dirname(__DIR__, 3) . '/vendor/autoload.php';
require \dirname(__DIR__, 3) . '/vendor/kubawerlos/php-cs-fixer-custom-fixers/bootstrap.php';
```

If I just comment out the first line the parallel runner starts working just fine. I'm not sure why does requiring the `autoload.php` file affect the runner? :thinking:

**Comment by keradus** [comment]
ideally, please share example repo to reproduce (you do not need to expose whole source code you have, just minimum reproducable example). For now, you are only one who can evidence and debug the issue.

https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/8014#issuecomment-2113527549

**Comment by Wirone** [comment]
> The weird thing is I've tried the parallel runner on another project (that one doesn't use `php-cs-fixer/shim` though, it uses `friendsofphp/php-cs-fixer`) and it's working just fine there so I'm really confused as to what could be the problem here.

This is something I noticed right after you reported the problem (using shim package), this should be verified in the first place. If I remember correctly I've been testing parallel runner using locally built PHAR file, but never tested shim installation.

**Comment by X-Coder264** [comment]
@keradus I've been trying to create a minimal reproducer but so far I was unable to do so.

All I can confirm for now is that if I remove `php-cs-fixer/shim` and require `friendsofphp/php-cs-fixer` instead then the parallel runner works fine.

**Comment by keradus** [comment]
I failed to reproduce with shim. if you cannot create minimal reproducer on your side, are you OK to grant temporary access to repo you experience the issue ?

**Comment by X-Coder264** [comment]
Sorry, I don't have the permissions on Github to give access to the project repository even if I wanted to.

**Comment by keradus** [comment]
are you able to create new repo with very limited content, that is allowing to reproduce issue you are facing ?

**Comment by sebastiaanluca** [comment]
Just chiming in here for what it's worth, but we're having the exact same issue. Works fine at the moment without parallel checks, but does nothing when enabled.

I can simplify our config to the following:

```
<?php

declare(strict_types=1);

use PhpCsFixer\Config;
use PhpCsFixer\Finder;
use PhpCsFixer\Runner\Parallel\ParallelConfigFactory;

$rules = [
    'blank_line_after_opening_tag' => true,
    'trailing_comma_in_multiline' => ['elements' => ['arrays', 'arguments', 'parameters']],
];

$finder = Finder::create()->in([__DIR__.'/config']);

return (new Config)
    ->setRules($rules)
    ->setFinder($finder)
    ->setUsingCache(false)
    ->setParallelConfig(ParallelConfigFactory::detect());
```

Which shows the progress bar but changes nothing to the files (for which I used the wrong code style for the 2 rules above):

```
PHP CS Fixer 3.57.1 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.7
Running analysis on 8 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "redacted/.php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
  0/61 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fixed 0 of 61 files in 0.104 seconds, 28.000 MB memory used
```

Will try to recreate a reproducible example later.

**Comment by enp-mrygiel** [comment]
For what it's worth - I'm facing the exact same issue.
Tested using 3.57.2

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/143383882/4648be4f-f545-4cfa-957e-22aec52e5399)

**Comment by Wirone** [comment]
@sebastiaanluca @enp-mrygiel how do you install Fixer? Shim package or something different?

**Comment by potsky** [comment]
Hi!
We have the same error but only when specifying a folder in command line:

- `vendor/bin/php-cs-fixer fix app/Models --config=.php-cs-fixer.dist.php -v --using-cache=no` fixes nothing
- `vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.dist.php -v --using-cache=no` works and fix files in `app/Models`

The output of the first command is:

```
vendor/bin/php-cs-fixer fix app/Models --config=.php-cs-fixer.dist.php -v --using-cache=no
PHP CS Fixer 3.57.1 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.7
Running analysis on 12 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from ".php-cs-fixer.dist.php".
Paths from configuration file have been overridden by paths provided as command arguments.
  0/62 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fixed 0 of 62 files in 0.109 seconds, 26.000 MB memory used
```

**Comment by Wirone** [comment]
Interesting, but I still can't reproduce it locally. When I run parallel analysis on Fixer's repo with a directory path as CLI argument, it's running correctly. Please guys, can anyone extract minimal reproducer repo so we can take a look what's happening there?

**Comment by potsky** [comment]
> Interesting, but I still can't reproduce it locally. When I run parallel analysis on Fixer's repo with a directory path as CLI argument, it's running correctly. Please guys, can anyone extract minimal reproducer repo so we can take a look what's happening there?

Just more context: in fact it works on Apple M1 with 10 CPUs and it fails on Apple M3 with 12 cores. Even if we reduce (on M3) the number of cores in docker to 10 or less, it always uses 12 cores.

Here is the output on an M1 which always works:

```
vendor/bin/php-cs-fixer fix app/Models --config=.php-cs-fixer.dist.php -v --using-cache=no
PHP CS Fixer 3.57.2 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.7
Running analysis on 10 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from ".php-cs-fixer.dist.php".
Paths from configuration file have been overridden by paths provided as command arguments.
 62/62 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) app/Models/Social.php (phpdoc_no_alias_tag, fully_qualified_strict_types, phpdoc_types_order, phpdoc_separation)
....
  52) app/Models/Group.php (phpdoc_no_alias_tag, fully_qualified_strict_types, phpdoc_types_order, phpdoc_separation)
```


Hope it can help...

**Comment by potsky** [comment]
Yes @Wirone we use `->setParallelConfig(PhpCsFixer\Runner\Parallel\ParallelConfigFactory::detect())` and everything is fine on my M1.
I will try to set the number of cores on a M3 and I will tell you.

**Comment by keradus** [comment]
@potsky , you think you can provide sth?
> Please guys, can anyone extract minimal reproducer repo so we can take a look what's happening there?

**Comment by potsky** [comment]
I have a M1 so I cannot provide anything directly, I will check tomorrow with my college @sylouuu if he can test some things and provide more information

**Comment by julienarcin** [comment]
I have the same problem, but it is related to the PHP version.

In my PHPStorm console with PHP 8.3.7 it works.

In my iTerm with PHP 8.2.19 it stops at 0%.

Same codebase, same shell (zsh), only the PHP version differs.

<img width="946" alt="Screenshot 2024-06-03 at 09 06 45" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/1089816/6d7c44db-9a6d-4784-adbf-4b59f52af674">

<img width="835" alt="Screenshot 2024-06-03 at 09 07 30" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/1089816/30439e98-f614-4359-a533-b31269f10588">

**Comment by julienarcin** [comment]
I have the same behavior.

The weird thing is that's it's working well inside PHPStorm terminal, but stuck at 0% when using iTerm2.

Both are using zsh and the same PHP version. And on the same machine (Macbook Pro Intel Core i9).

It doesn't seem hardware-related, at least on my side. 🤷‍♂️

**Comment by potsky** [comment]
> I have the same behavior.
> 
> The weird thing is that's it's working well inside PHPStorm terminal, but stuck at 0% when using iTerm2.
> 
> Both with the same codebase, with zsh and the same PHP version. And on the same machine (Macbook Pro Intel Core i9).
> 
> It doesn't seem hardware-related, at least on my side. 🤷‍♂️

Really interesting 🤔 

It works on my Mac M1 with iTerm2 terminal.
It falls on the Mac M3 with Warp terminal of my colleague.

**Comment by Wirone** [comment]
[In our CI it works](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/9347558766) on PHP 7.4-8.4, on Linux, MacOS and Windows. Locally I use parallel runner on MacOS (M1) within iTerm2 and PHPStorm's terminal (the new one, but also the old one which just spawns the same ZSH session as iTerm2). Tested also with built-in Terminal (v2.14), also works properly. It works for files configured in the config file, but also when I set file/dir as a CLI command argument... I don't have any reproducer currently and can't help 🙁.

**Comment by EdenMl** [comment]
Hello,
I am currently experiencing the same problem after upgrading to **version 3.57.0**.

On my MacOS with an Intel Core i5 chip, my files are no longer being verified, and I receive the message _"Fixed 0 of 4523 files in 0.138 seconds."_.

In the meantime, I've removed the `setParallelConfig(ParallelConfigFactory::detect())` method from the **_.php-cs-fixer.php_** file.

**Comment by enp-mrygiel** [comment]
bin/php-cs-fixer fix --dry-run
PHP CS Fixer 3.58.1 (8fdc3d5) 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.1.28
Running analysis on 4 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "/var/www/.php-cs-fixer.php".
Using cache file ".php-cs-fixer.cache".
     0/65958 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Found 0 of 65958 files that can be fixed in 0.418 seconds, 564.01 MB memory used

problem still exists in latest 3.58.1 version

**Comment by MazzMazz** [comment]
For me this is still happening in Version `3.59.3`

```
> vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.php --verbose
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.0
Running analysis on 12 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from ".php-cs-fixer.php".
    0/2066 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fixed 0 of 2066 files in 0.111 seconds, 24.00 MB memory used
```

**Comment by MazzMazz** [comment]
tl;dr: 
I removed the PHP installed via HERD and installed PHP 8.3 via BREW. And it works 🥳 
I don't know what you do with this information but maybe it helps

---

I created a Repo in which the problem exists (at least for me) on my Mac M2 Pro. It is with the default Laravel Project. 
I guess this repo is working on your machine - but I don't know how i could support otherwise.

https://gitlab.com/MazzMazz/testrepocsfixer (using `friendsofphp/php-cs-fixer`) 

- It is the same result in PHPStorm Terminal and default Mac Terminal (zsh)
- I tested project namings (testRepo, testrepo, test-repo, test repo) but couldn't get any different results
- I tried with `sudo` - no difference
- I tried the setting with autoDetect, 2 cores (like in the repo) but also 60 (even though i don't have 60 cores). No difference

```
vendor/bin/php-cs-fixer fix --dry-run -vvv
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.0
Running analysis on 2 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "/Users/myuser/PhpstormProjects/testrepo/.php-cs-fixer.php".
 0/3 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Found 0 of 3 files that can be fixed in 0.110 seconds, 20.00 MB memory used
```

Composer version 2.7.7 2024-06-10 22:11:12
PHP version 8.3.0 (via Herd)

Updated PHP to Version 8.3.9

```
vendor/bin/php-cs-fixer fix --dry-run -vvv
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.9
Running analysis on 2 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "/Users/myuser/PhpstormProjects/testrepo/.php-cs-fixer.php".
 0/3 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Found 0 of 3 files that can be fixed in 0.109 seconds, 20.00 MB memory used
```

Downgraded cs-fixer to 3.57.0 - same result
Tried with PHP 8.2 - same result

```
PHP CS Fixer 3.57.0 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.2.14
Running analysis on 2 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "/Users/myuser/PhpstormProjects/testrepo/.php-cs-fixer.php".
 0/3 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Found 0 of 3 files that can be fixed in 0.109 seconds, 18.000 MB memory used
```

I removed the PHP installed via HERD and installed PHP 8.3 via BREW. And it works 🥳 
I don't know what you do with this information but maybe it helps

```
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.9
Running analysis on 2 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "/Users/mathiasmaier/PhpstormProjects/testrepo/.php-cs-fixer.php".
 3/3 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) app/Models/User.php (single_trait_insert_per_statement, no_multiple_statements_per_line)

Found 1 of 3 files that can be fixed in 0.106 seconds, 20.00 MB memory used
```

**Comment by Wirone** [comment]
@MazzMazz thank you for providing this context! Could you please check `php -m` on Herd (and Brew version, for comparison)?

**Comment by MazzMazz** [comment]
PHP via Herd

```
[PHP Modules]
bcmath
bz2
calendar
Core
ctype
curl
date
dba
dom
exif
FFI
fileinfo
filter
ftp
gd
gmp
hash
iconv
igbinary
imagick
imap
intl
json
ldap
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_pgsql
pdo_sqlite
pgsql
Phar
posix
random
readline
redis
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib
zstd

[Zend Modules]
Zend OPcache

````

PHP via Brew

```
[PHP Modules]
bcmath
bz2
calendar
Core
ctype
curl
date
dba
dom
exif
FFI
fileinfo
filter
ftp
gd
gettext
gmp
hash
iconv
intl
json
ldap
libxml
mbstring
mysqli
mysqlnd
odbc
openssl
pcntl
pcre
PDO
pdo_dblib
pdo_mysql
PDO_ODBC
pdo_pgsql
pdo_sqlite
pgsql
Phar
posix
pspell
random
readline
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tidy
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache
```

**Comment by MazzMazz** [comment]
@Wirone have you had any luck with this information?

**Comment by Wirone** [comment]
@MazzMazz unfortunately did not have time to debug it yet.

**Comment by keradus** [comment]
Hey everyone.
Before writing "I experience it too!", please help to reproduce this outside of your local machine. We understand it is happening, yet without being able to reproduce outside of your local machine, we won't be able to provide the fix.

Ideally, provide an example repository with minimal content (config file, Fixer version, command to run, example file(s) that are expected to be processed).
[example](https://github.com/keradus/PHP-CS-Fixer-repro-cases/tree/6715)

**Comment by mateusjatenee** [comment]
FWIW, I also managed to fix the problem by removing using Homebrew's PHP binary instead of Herd's. 

Specs:
Macbook Pro M3 Max
Doesn't work on Herd's 8.2 or 8.3
Works well with Homebrew @ 8.3

I'm crunched on time right now, but I'll try digging into it after Laracon and see if I can help.

Here's a diff between Homebrew's `php -m` and Herd's: https://www.diffchecker.com/cx95xpJJ/

**Comment by BertvanHoekelen** [comment]
I've been having the same problem since the beginning and I'm also running herd (so some having "I experience this too" comments). Have helped me pinpointing the issue.

It's kind of a herd problem, as I have been running into the same issues with the node binary. The problem is that herd is running in `/Library/Application Support/Herd`, having a space in the binary location is causing many issues.  Changing the php [binary](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/Runner/Parallel/ProcessFactory.php#L58) here 

from:

```php
$phpBinary = (new PhpExecutableFinder())->find(false);
```

into

```php|
$phpBinary = (new PhpExecutableFinder())->find(false);
$phpBinary = str_replace(" ", "\\ ", $phpBinary);
```

solves the problem with running php-cs-fixer in parallel mode. But I think the problem real problem is in Symfony's [process library](https://github.com/symfony/process/blob/7.1/PhpExecutableFinder.php). 

I'm happy to do PR there, but I'm not entirely sure they are open to fixes there. They have their issues tab closed and looking at the PR's they are all denied. So I've done it [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8180)

**Comment by Wirone** [comment]
For anyone of you that encountered this issue - please verify if it works for you after #8180 is released, and let us know if there are any other problems related to files not getting fixed.

**Comment by keradus** [comment]
thanks!

released under 3.63+
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/releases/tag/v3.63.0

if sth not working still, I encourage to raise a new issue ;)

**Comment by X-Coder264** [comment]
I still have the same issue (it still dies on 0% without fixing any file).

```
PHP CS Fixer 3.63.1 (2af4b11) 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.2.18
Running analysis on 20 cores with 10 files per process.
Parallel runner is an experimental feature and may be unstable, use it at your own risk. Feedback highly appreciated!
Loaded config default from "./app/tools/php-cs-fixer/config.php".
     0/12109 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fixed 0 of 12109 files in 1.038 seconds, 124.00 MB memory used
```

**Comment by Wirone** [comment]
Seems like the fix resolved only _some_ cases, probably shim package usage needs to be investigated more.
